pkg/imgpkg/imagetar/tar_reader.go

```
@@ -33,101 +39,132 @@ func (r TarReader) Read() ([]imagedesc.ImageOrIndex, error) {
+		}
+	}
+
+	var result []v1.Layer
+	for _, layer := range layerCache {
+		if layer != nil {
+			result = append(result, layer)
 		}
 	}
 
 	return result, nil
 }
 
-func (r TarReader) presentLayersForImage(img v1.Image) ([]v1.Layer, error) {
-	var result []v1.Layer
+func (r *TarReader) presentLayersForImage(img v1.Image, layerCache map[string]v1.Layer) error {
 	layers, err := img.Layers()
 	if err != nil {
-		return nil, fmt.Errorf("Unable to retrieve layers: %s", err)
+		return fmt.Errorf("Unable to retrieve layers: %s", err)
 	}
 
+	errChan := make(chan error, len(layers))
 	for _, layer := range layers {
 		h, err := layer.Digest()
 		if err != nil {
-			return nil, fmt.Errorf("Unable to get digest from layer: %s", err)
+			return fmt.Errorf("Unable to get digest from layer: %s", err)
 		}
-		r, err := layer.Compressed()
-		if err != nil {
+		r.lock.Lock()
+		if _, ok := layerCache[h.String()]; ok {
+			errChan <- nil
+			r.lock.Unlock()
 			continue
 		}
+		layerCache[h.String()] = nil
+		r.lock.Unlock()
 
-		size, err := layer.Size()
-		if err != nil {
-			return nil, err
-		}
-		closer, err := verify.ReadCloser(r, size, h)
-		if err != nil {
-			return nil, err
-		}
+		go func() {
+			layer := layer
+			h := h
 
-		_, err = io.Copy(io.Discard, closer)
-		if err != nil {
-			continue
-		}
+			reader, err := layer.Compressed()
+			if err != nil {
+				errChan <- nil
+				return
+			}
+
+			size, err := layer.Size()
+			if err != nil {
+				errChan <- err
+				return
+			}
+			closer, err := verify.ReadCloser(reader, size, h)
+			if err != nil {
+				errChan <- err
+				return
+			}
+
+			r.throttle.Take()
+			_, err = io.Copy(io.Discard, closer)
+			r.throttle.Done()
+			if err != nil {
+				errChan <- nil
+				return
+			}
 
-		result = append(result, layer)
+			r.lock.Lock()
+			layerCache[h.String()] = layer
+			r.lock.Unlock()
+			errChan <- nil
+		}()
 	}
-	return result, nil
+
+	for range layers {
+		if err := <-errChan; err != nil {
+			return err
```

Is there any value in aggregating errors in case of multiple ones here?

---

we are failing fast here but I'm not sure if it is worth it or not to wait for all the layers.